### PR TITLE
Deletion

### PR DESCRIPTION
**Pull request by Lewington-pitsos**
add functions to delete gates and update gate names, also clean up some of the testing code and squash some bugs

### CODE REVIEW

File: contracts/PaymentReceiver.sol [hunk 1]
```diff
@@ -24,23 +24,35 @@ contract PaymentReceiver is IPaymentReceiver, Database {
         host = _host;
         cfa = _cfa;
     }
- 
+    
+    function deleteGate(uint256 _gateId) public override {
```
Comment (2manslkh) [inline contracts/PaymentReceiver.sol:28]: Make sure only owner of the gate can rename

File: contracts/Database.sol [hunk 2]
```diff
@@ -25,14 +25,22 @@ contract Database {
         onlyToken = _token;
     }
 
+    function renameGate(uint256 _gateId, string memory _name) external {
```
Comment (2manslkh) [inline contracts/Database.sol:28]: Make sure only the owner of the gate can rename the gate name

File: contracts/Database.sol [hunk 3]
```diff
@@ -43,9 +51,41 @@ contract Database {
...
+    function _checkGateName(string memory _name) private view {
+        bytes32 _nameId = keccak256(bytes(_name));
+        uint256[] memory gateIds = addressGates[msg.sender];
+        for (uint256 index = 0; index < gateIds.length; index++) {
+            Gate memory _gate = gates[gateIds[index]];
+            require(keccak256(bytes(_gate.name)) != _nameId, string(abi.encodePacked(_name, "is already taken by another gate")));
+        }
+    }
+
+    function deleteGate(uint256 _gateId) public virtual {
```
Comment (2manslkh) [inline contracts/Database.sol:65]: Make sure only the owner of the gate can delete the gate

**Review by 2manslkh — CHANGES_REQUESTED** [review]
I think if PaymentReciever.sol is going to inherit database, Database.sol should be an abstract contract

File: contracts/Database.sol [hunk 4]
```diff
@@ -43,9 +51,41 @@ contract Database {
         return _id;
     }
 
+    function _checkGateName(string memory _name) private view {
```
Comment (oschleic) [inline contracts/Database.sol:56]: Also change this to a mapping(string => uint256) and just check if it exists

Comment (Lewington-pitsos) [reply contracts/Database.sol:56]: as in, keep a new gate name mapping for each merchant? 

Is the extra storage less expensive than the extra computation in this case?

File: contracts/Database.sol (hunk 3 above)
Comment (Lewington-pitsos) [reply contracts/Database.sol:65]: this is already the case but I have added a test to ensure that it works

**Pull request merged**